source/windows/thread.c

```diff
@@ -46,38 +57,49 @@ int aws_thread_create(struct aws_thread *thread, void(*func)(void *arg), void *c
 
     struct thread_wrapper *thread_wrapper = (struct thread_wrapper *)aws_mem_acquire(thread->allocator, sizeof(struct thread_wrapper));
     thread_wrapper->allocator = thread->allocator;
-    thread_wrapper->arg = context;
+    thread_wrapper->arg = arg;
     thread_wrapper->func = func;
     thread->thread_handle = CreateThread(0, stack_size, thread_wrapper_fn, (LPVOID)thread_wrapper, 0, &thread->thread_id);
 
     if (!thread->thread_handle) {
         return aws_raise_error(AWS_ERROR_THREAD_INSUFFICIENT_RESOURCE);
     }
 
+    thread->detach_state = AWS_THREAD_JOINABLE;
     return AWS_OP_SUCCESS;
 }
 
 uint64_t aws_thread_get_id(struct aws_thread *thread) {
     return thread->thread_id;
 }
 
+aws_thread_detach_state aws_thread_get_detach_state(struct aws_thread *thread) {
+    return thread->detach_state;
+}
+
 int aws_thread_detach(struct aws_thread *thread) {
 
     if (thread->thread_handle) {
+        thread->detach_state = AWS_THREAD_DETACHED;
         CloseHandle(thread->thread_handle);
         thread->thread_handle = 0;
     }
     return AWS_OP_SUCCESS;
 }
 
 int aws_thread_join(struct aws_thread *thread) {
+    if (thread->detach_state == AWS_THREAD_JOINABLE) {
+        WaitForSingleObject(thread->thread_handle, INFINITE);
+        thread->detach_state = AWS_THREAD_JOIN_COMPLETED;
+        return AWS_OP_SUCCESS;
+    }
 
-    WaitForSingleObject(thread->thread_handle, INFINITE);
-
-    return AWS_OP_SUCCESS;
+    return aws_raise_error(AWS_ERROR_THREAD_NOT_JOINABLE);
 }
 
 void aws_thread_clean_up(struct aws_thread *thread) {
+    /* cleaning up a joinable thread is undefined. Don't do it. Either join or detach. The cpp stdlib actually calls std::terminate in this case */
```

Comment:
Whatever the C++ stdlib does is irrelevant to our API. Why shouldn't we allow you to clean up a joinable thread? We clearly know if it's joinable or not, and we know how to detach it. And, in fact, if we remove this assert it'll work just fine. What future-proofing are we trying to accomplish here?

Reply:
the assert is because if you aren't either calling join or detach before cleaning up the thread it's a bug and quite possibly a resource leak.

Reply:
I don't buy that argument. The idea that 'join' releases resources is a very pthread-specific notion. In windows, waiting for a handle does not release it. In java, `Thread.join` does not release any resources.

I would argue that `detach` is superfluous (why not just use `clean_up`?) and `join` should be idempotent.

Reply:
ok, deleting the "detached" state from the thread, and deleting the detach function, moving to clean up.